File: src/main/java/gay/skitbet/jackiro/JackiroLauncher.java
package gay.skitbet.jackiro;

import com.google.gson.Gson;
import com.google.gson.GsonBuilder;
import gay.skitbet.jackiro.utils.JackiroConfig;
import org.apache.commons.io.FileUtils;

import java.io.File;
import java.io.IOException;
import java.nio.charset.StandardCharsets;
import java.util.Collections;

public class JackiroLauncher {

    private static final Gson GSON = new GsonBuilder().setPrettyPrinting().create();
    private static final File CONFIG_FILE = new File("./config.json");

    public static void main(String[] args) {
        JackiroConfig config = loadOrCreateConfig();
        Jackiro.initialize(config);
    }

    /**
     * Loads or creates a configuration for the bot!
     * @return
     */
    private static JackiroConfig loadOrCreateConfig() {
        if (CONFIG_FILE.exists()) {
            try {
                String json = FileUtils.readFileToString(CONFIG_FILE, StandardCharsets.UTF_8);
                return GSON.fromJson(json, JackiroConfig.class);
            } catch (IOException e) {
                System.err.println("Failed to read config file: " + e.getMessage());
                e.printStackTrace();
                System.exit(1);
            }
        } else {
            createDefaultConfig();
            System.out.println("\nConfig created! Please configure the `config.json` before restarting the bot.");
            System.exit(1);
        }
        return null;
    }

    /**
     * Creates default config for the bot
     */
    private static void createDefaultConfig() {
        System.out.println("Jackiro Bot - First time setup");
        System.out.println("Creating a default config.json...");

        JackiroConfig defaultConfig = new JackiroConfig()
                .setClientSecret("client secret here")
                .setClientID("client ID here")
                .setClientToken("bot token here")
                .setOwnerId("Discord user ID here")
                .setCurrentlyPlaying(Collections.singletonList("Life"));

        try {
            FileUtils.writeStringToFile(CONFIG_FILE, GSON.toJson(defaultConfig), StandardCharsets.UTF_8);
        } catch (IOException e) {
            System.err.println("Failed to create default config: " + e.getMessage());
            e.printStackTrace();
        }
    }
}

File: src/main/kotlin/gay/skitbet/jackiro/JackiroLauncher.kt
package gay.skitbet.jackiro

import com.google.gson.GsonBuilder
import gay.skitbet.jackiro.utils.JackiroConfig
import org.apache.commons.io.FileUtils
import java.io.File
import java.nio.charset.StandardCharsets

object JackiroLauncher {

    private val gson = GsonBuilder().setPrettyPrinting().create()
    private val configFile = File("./config.json")

    @JvmStatic
    fun main(args: Array<String>) {
        val config = loadOrCreateConfig()
        Jackiro.initialize(config)
    }

    /**
     * Loads or creates a configuration for the bot!
     */
    private fun loadOrCreateConfig(): JackiroConfig {
        return if (configFile.exists()) {
            try {
                val json = FileUtils.readFileToString(configFile, StandardCharsets.UTF_8 as String?)
                gson.fromJson(json, JackiroConfig::class.java)
            } catch (e: Exception) {
                System.err.println("Failed to read config file: ${e.message}")
                e.printStackTrace()
                System.exit(1)
                throw RuntimeException("Unreachable") // for compiler
            }
        } else {
            createDefaultConfig()
            println("\nConfig created! Please configure the `config.json` before restarting the bot.")
            System.exit(1)
            throw RuntimeException("Unreachable") // for compiler
        }
    }

    /**
     * Creates default config for the bot
     */
    private fun createDefaultConfig() {
        println("Jackiro Bot - First time setup")
        println("Creating a default config.json...")

        val defaultConfig = JackiroConfig(
            clientSecret = "client secret here",
            clientID = "client ID here",
            clientToken = "bot token here",
            ownerId = "Discord user ID here",
            currentlyPlaying = mutableListOf("Life")
        )

        try {
            FileUtils.writeStringToFile(configFile, gson.toJson(defaultConfig), StandardCharsets.UTF_8 as String?)
        } catch (e: Exception) {
            System.err.println("Failed to create default config: ${e.message}")
            e.printStackTrace()
        }
    }
}
